app/models/data_set.rb

```diff
@@ -101,31 +97,52 @@ def processing_complete?
   end
 
   def data_file=(file)
-    self.csv_data = read_as_utf8(file)
-
-    @need_csv_processing = true
+    if file.nil?
+      @csv_data = nil
+      @need_csv_processing = false
+    else
+      @csv_data = CsvData.new(data_file: file)
+      @need_csv_processing = true
+    end
   end
 
   def schedule_csv_processing
     if @need_csv_processing
+      @csv_data.save!
       ProcessCsvDataWorker.perform_async(service.id.to_s, self.version)
-      @need_csv_processing = false
+      @need_csv_processing = nil
     end
   end
 
+  def csv_data
+    @csv_data ||= CsvData.where(service_slug: service.slug, data_set_version: self.version).first
+  end
+
+  def csv_data_is_valid
+    return if @csv_data.nil? || @csv_data.destroyed?
+    @csv_data.service_slug = service.slug
+    @csv_data.data_set_version = self.version
```

Comment:
Because they might not exist yet.  Consider `DataSet.new(data_file: foo, service: bar)` - version definitely hasn't been set (it won't be until set_version is called as part of the validation cycle) and `service` might not be set yet depending on the order that the attributes are set during `initialize` (e.g. if `service` is set before `data_file` then it will be available in `data_file=`, but if `service` is set after `data_file` it won't be available in `data_file=`).  Originally they were set in that method, but it broke because of this. :(